### DISCUSSION TITLE
Making MvvmDialogs Trimming-Friendly

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#10495] [Q&A]
With [MvvmDialogs](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/), views are created via reflection.

This is a problem with assembly trimming, as the views are never directly referenced.

What's the best solution to make apps with MvvmDialogs trimming-friendly?

One solution I tried is to inherit all my windows from CommonWindow

```c#
[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]
public abstract class CommonWindow : Window
```

Unfortunately, that did not work.

> System.MissingMethodException: Cannot dynamically create an instance of type 'HanumanInstitute.Converter432hz.Views.MainView'. Reason: No parameterless constructor defined.

As a work-around, I can add this to my app startup and it works. There's got to be a better solution though, and I have to document the proper solution in the MvvmDialogs repo.

    var _ = new Window[] { new MainView(), new AboutView(), new SettingsView(), new AskFileActionView() };

What would be the best way to handle this?

Edit: Or maybe... add a function to add all Views manually into MvvmDialogs and remove the need for reflection altogether.

**Comment by maxkatz6** [comment]
> This is a problem with assembly trimming, as the views are never directly referenced.

I assume, it's not something that you can pre-register in the library code (otherwise it could be easily solved with [ILLink.Descriptors.xml](https://github.com/AvaloniaUI/Avalonia/pull/9479)).

In general, using reflection to instantiate views wouldn't be as effective, as factories. 
Type needs to be somehow rooted in the assembly, used somewhere (including constructors).
Users can define [TrimmerRootAssembly or TrimmerRootDescriptor](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0) to define types/assemblies that needs to be saved from being trimmed.
Note though, if whole assembly is rooted, then trimming will be limited on it.

Is it possible to register view factories?
I.e. `factories.Add(key: typeof(MainViewModel), value: () => new MainView())` or with some extensions `.RegisterDialog<MainViewModel, MainView>()` and `.RegisterDialog<MainViewModel, MainView>(() => new MainView())` ?

Alternatively, you can write a source generator that will be shipped with the main nuget package.
In the source generator you can lookup view models in the assembly, find matched views, and generate partial method or class code. But even then users will need to register this generated method/class in the library.
Like these https://github.com/AvaloniaUI/Avalonia/pull/7935 or https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/

Microsoft have quite a lot of ways to make libraries trimmable friendly. But I wouldn't say there is any easy for both library's developers and users.

**Comment by mysteryx93** [comment]
I was thinking of adding

    dialogService.Register<MainView>().Register<MainWindow>();

or even

    dialogService.Register<MainView, MainWindow>();

and then looking at registered views before relying on reflection

but while being there; might as well register the View with the ViewModel directly and remove the need for the name Locator altogether

    dialogService.Register<MainViewModel, MainView>();

but what about when it can be View or Window depending on platform?

    dialogService.Register<MainViewModel, MainView, MainWindow>();

and someone might have the need for more than 2 view types for whatever reason

    dialogService.Register<MainViewModel, MainView, MainWindow, MainViewWeb>();

Is referencing the type sufficient (it would still be created via reflection), or do I need to explicitly call the constructor?

I'll also note that all Dependency Injection Containers contains face a similar situation?

How about this as standard method

    dialogService.Register<MainViewModel>(Type[] views);

And the rest as extension methods to provide type checking and allowing for custom extensions

    dialogService.Register<MainViewModel, MainView>(); // where MainView is Window
    dialogService.Register<MainViewModel, MainView, MainWindow>(); // where MainView is UserControl, MainWindow is Window

**Reply by maxkatz6** [reply]
> I'll also note that all Dependency Injection Containers contains face a similar situation?

Kinda. Mostly because they have set proper trimming attributes, so minimal reflection can work.
But I haven't followed these developments that much.
https://github.com/dotnet/runtime/issues/44432
Reflection-free mode is another story - https://codevision.medium.com/dependency-injection-for-native-aot-e6cc90bef395 - but it's more like an experimental feature which probably won't be common in a long time.

**Reply by maxkatz6** [reply]
> dialogService.Register<MainViewModel>(Type[] views);

I would try that, yes. Maybe define it like:
```
void Register<TViewModel>([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] params Type[] views)
```
This code should work with single type. But no idea if it will work with array of types.
If it won't, you can hardcode multiple register methods with 1, 2 and 3 type options (if not more).

**Reply by maxkatz6** [reply]
[DynamicallyAccessedMembers] would tell the compiler, that if any type is passed as the parameter, then compiler should keep constructors of this type from being trimmed.

**Reply by mysteryx93** [reply]
In the case of `Type[] views`, it might apply that to the constructor of Array class. Would have to try.

btw read some of the articles. Pure.DI looks like an interesting DI container, do you know anything about that in comparison to Splat which I'm currently using? Reactive programmer says he needs to finish the Source Generator for Splat; but it seems Pure.DI has already done more than what Splat wants to do?

**Reply by mysteryx93** [reply]
Must have been tired yesterday. Confused Splat with PropertyChanged.Fody, where the author said he wanted to finish a Source Generator for it to replace Fody. Splat is already a Source Generator. Pure.DI and Splat actually look very similar to each other.

I already have a design where I can pass a custom factory to resolve types. I probably can achieve this by writing a new factory type with 0 design change.

**Comment by mysteryx93** [comment]
The good news is that `StrongViewLocator` is now implemented for Avalonia.

The bad news is that it's not yet solving the trimming problem.

    public StrongViewLocator Register<TViewModel>([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type view)
        where TViewModel : INotifyPropertyChanged
    {
        RegisterBase<TViewModel>(view);
        return this;
    }

I have various extension methods that eventually call this (with the Desktop or Navigation view).

I then configure it like this.

    var viewLocator = new StrongViewLocator() { ForceSinglePageNavigation = false }
        .Register<MainViewModel, MainView, MainWindow>()
        .Register<CurrentTimeViewModel, CurrentTimeView, CurrentTimeWindow>()
        .Register<ConfirmCloseViewModel, ConfirmCloseView, ConfirmCloseWindow>();

In this context, is it applying `DynamicallyAccessedMemberTypes.PublicConstructors` to the `Type` class?

My approach with MvvmDialogs is that it doesn't matter how much black wizardry I do in the back-end; I want the front-end API to be nice and simple. But I'd rather avoid the Source Generator route if possible...

Any idea here?

**Reply by mysteryx93** [reply]
The problem is that I rely heavily on Activator.CreateInstance. Avoiding reflection to determine views but still using reflection to create type does feel like a half-baked job. It would be nice to support reflection-free mode.

That however require pretty deep refactoring... and the code may need both the view type and view creation function (need to know the return type without calling the creation method), which would be ugly to configure. That would then need to be combined with Source Generation........

It seems that's what it takes to support
- Assembly trimming
- Reflection-free mode
- Clean user syntax

Black wizardry in the back-end, and the front-end user has no clue about it.

**Reply by mysteryx93** [reply]
Great, refactored to truly be Reflection-free, but the definition is ugly.

```c#
ViewLocator = new StrongViewLocator()
    .Register<MainWindowViewModel>(new ViewDefinition(typeof(MainWindow), () => new MainWindow()))
    .Register<AddTextDialogViewModel>(new ViewDefinition(typeof(AddTextDialog), () => new AddTextDialog()));
```

Will need a Source Generator to turn this into the code above.

```c#
ViewLocator = new StrongViewLocator()
    .Register<MainWindowViewModel, MainWindow>()
    .Register<AddTextDialogViewModel, AddTextDialog>();
```
